compiler/src/iree/compiler/Codegen/Common/VectorizePackUnPackOps.cpp

```diff
@@ -0,0 +1,152 @@
+                return tileSizes;
+              }));
+      funcOp->walk([&](tensor::PackOp op) {
+        FailureOr<scf::SCFTileAndFuseResult> tileAndFuseResult =
+            scf::tileConsumerAndFuseProducerGreedilyUsingSCFForOp(
+                rewriter, cast<TilingInterface>(op.getOperation()),
+                packOptions);
+        if (failed(tileAndFuseResult)) return signalPassFailure();
+        rewriter.replaceOp(op, tileAndFuseResult->replacements[op.getResult()]);
+      });
+
+      auto unpackTilingOptions =
+          scf::SCFTilingOptions().setTileSizeComputationFunction(
+              [](OpBuilder &builder, Operation *op) {
+                Location loc = op->getLoc();
+                auto unpackOp = cast<tensor::UnPackOp>(op);
+                int numLoops = unpackOp.getDestRank();
+                auto dimAndTileMapping = unpackOp.getDimAndTileMapping();
+                SmallVector<Value> tileSizes;
+                for (int i = 0; i < numLoops; ++i) {
+                  if (dimAndTileMapping.count(i)) {
+                    tileSizes.push_back(getValueOrCreateConstantIndexOp(
+                        builder, loc, dimAndTileMapping[i]));
+                  } else {
+                    tileSizes.push_back(builder.createOrFold<tensor::DimOp>(
+                        loc, unpackOp.getDest(), i));
+                  }
+                }
+                return tileSizes;
+              });
+      funcOp->walk([&](tensor::UnPackOp op) {
+        FailureOr<scf::SCFTilingResult> tilingResult = scf::tileUsingSCFForOp(
+            rewriter, cast<TilingInterface>(op.getOperation()),
+            unpackTilingOptions);
+        if (failed(tilingResult)) return signalPassFailure();
+        rewriter.replaceOp(op, tilingResult->replacements);
+      });
+    }
+
+    LLVM_DEBUG({
+      llvm::dbgs()
+          << "--- After applying tiling that makes outer dims be all 1s ---\n";
+      funcOp.print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
+      llvm::dbgs() << "\n\n";
+    });
+
+    // Generalize pack and unpack ops and canonicalize tiled ops.
+    {
+      RewritePatternSet patterns(ctx);
+      linalg::populateLinalgTilingCanonicalizationPatterns(patterns);
+      patterns.add<linalg::GeneralizeOuterUnitDimsPackOpPattern,
+                   linalg::GeneralizeOuterUnitDimsUnPackOpPattern>(ctx);
+      if (failed(applyPatternsAndFoldGreedily(getOperation(),
+                                              std::move(patterns)))) {
+        return signalPassFailure();
+      }
+    }
+
+    LLVM_DEBUG({
+      llvm::dbgs()
+          << "--- After generalizing tensor.pack and tensor.unpack ops ---\n";
+      funcOp.print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
+      llvm::dbgs() << "\n\n";
+    });
+
+    // Kick in generic vectorizer.
+    {
+      RewritePatternSet patterns(ctx);
+      patterns.add<IREE::LinalgExt::LinalgVectorizationPattern>(ctx);
```

Comment:
Is all of this available in main too?

Reply:
The implementation is all in MLIR repo. This patterns is just a wrapper for calling `vectorize` method.

https://github.com/iree-org/iree/blob/234f4b2f0f25d693e6e488908d5c5232851346af/llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/Transforms.cpp#L258-L268